File: code/diggerconfig.py
import os
import yaml

try:
    from yaml import CLoader as Loader, CDumper as Dumper
except ImportError:
    from yaml import Loader, Dumper


class DiggerConfig:
    def __init__(self):
        if os.path.exists("digger.yml"):
            with open("digger.yml", "r") as f:
                self.config = yaml.load(f, Loader=Loader)
        else:
            self.config = {"projects": {"name": "default", "dir": ""}}

    def get_project(self, project_name):
        for project in self.config["projects"]:
            if project_name == project["name"]:
                return project
        return None

    def get_projects(self, project_name=None):
        if self.config and self.config.get("projects") and not project_name:
            return self.config["projects"]
        elif project_name:
            project = self.get_project(project_name)
            if project:
                return [project]
        return []

    def get_modified_projects(self, changed_files):
        all_projects = self.get_projects()
        result = []

        for p in all_projects:
            print(f'p: {p}')
            for f in changed_files:
                if "dir" in p and f.filename.startswith(p["dir"]):
                    result.append(p)
        return result

    def get_directory(self, project_name):
        project = self.get_project(project_name)
        return project.get("dir", None)


digger_config = DiggerConfig()

File: digger_config.go
package main

import (
	"fmt"
	"gopkg.in/yaml.v2"
	"os"
)

type DiggerConfig struct {
	Projects []Project `yaml:"projects"`
}

type Project struct {
	Name string `yaml:"name"`
	Dir  string `yaml:"dir"`
}

func NewDiggerConfig() (*DiggerConfig, error) {
	config := &DiggerConfig{}
	if data, err := os.ReadFile("digger.yml"); err == nil {
		if err := yaml.Unmarshal(data, config); err != nil {
			return nil, fmt.Errorf("error parsing digger.yml: %v", err)
		}
	} else {
		config.Projects = make([]Project, 1)
		config.Projects[0] = Project{Name: "default", Dir: "."}
		return config, nil
	}
	return config, nil
}

func (c *DiggerConfig) GetProject(projectName string) *Project {
	for _, project := range c.Projects {
		if projectName == project.Name {
			return &project
		}
	}
	return nil
}

func (c *DiggerConfig) GetProjects(projectName string) []Project {
	if projectName == "" {
		return c.Projects
	}
	project := c.GetProject(projectName)
	if project == nil {
		return nil
	}
	return []Project{*project}
}

func (c *DiggerConfig) GetModifiedProjects(changedFiles []string) []Project {
	var result []Project
	for _, project := range c.Projects {
		for _, file := range changedFiles {
			if project.Dir != "" && file[:len(project.Dir)] == project.Dir {
				result = append(result, project)
				break
			}
		}
	}
	return result
}

func (c *DiggerConfig) GetDirectory(projectName string) string {
	project := c.GetProject(projectName)
	if project == nil {
		return ""
	}
	return project.Dir
}

type File struct {
	Filename string
}
